File: c++/RayTracer.cpp
#include <cmath>
#include <vector>
#include <iostream>
#include <memory>
#include <fstream>
#include <chrono>
#include <optional>

const double FarAway = 1000000.0;
using UInt8 = unsigned char;

struct RgbColor
{
    UInt8 b, g, r, a;
};

struct Vector
{
    double x, y, z;

    Vector() : x(0), y(0), z(0) {}

    Vector(double x, double y, double z) : x(x), y(y), z(z) { }

    double Length() const
    {
        return sqrt(x * x + y * y + z * z);
    }

    Vector Norm() const
    {
        double mag = Length();
        double div = (mag == 0) ? FarAway : 1.0 / mag;
        return *this * div;
    }

    Vector Cross(const Vector& v) const
    {
        return Vector(
            y * v.z - z * v.y,
            z * v.x - x * v.z,
            x * v.y - y * v.x
        );
    }

    Vector operator*(double k) const
    {
        return Vector(k * x, k * y, k * z);
    }

    double operator*(const Vector& v) const
    {
        return x * v.x + y * v.y + z * v.z;
    }

    Vector operator+(const Vector& v) const
    {
        return Vector(x + v.x, y + v.y, z + v.z);
    }

    Vector operator-(const Vector& v) const
    {
        return Vector(x - v.x, y - v.y, z - v.z);
    }
};

struct Color
{
    double r, g, b;

    static Color White;
    static Color Grey;
    static Color Black;
    static Color Background;
    static Color DefaultColor;

    Color() : r(0.0), g(0.0), b(0.0) {}

    Color(double r, double g, double b) : r(r), g(g), b(b) { }

    Color Scale(double k) const
    {
        return Color(k * r, k * g, k * b);
    }

    Color operator * (const Color& c) const
    {
        return Color(r * c.r, g * c.g, b * c.b);
    }

    Color operator + (const Color& c) const
    {
        return Color(r + c.r, g + c.g, b + c.b);
    }

    RgbColor ToDrawingColor() const
    {
        return RgbColor{ Clamp(b), Clamp(g), Clamp(r), 255 };
    }

    static UInt8 Clamp(double c)
    {
        int x = (int)(c * 255);
        if (x < 0)   x = 0;
        if (x > 255) x = 255;
        return (UInt8)x;
    }
};

Color Color::White = Color(1.0, 1.0, 1.0);
Color Color::Grey = Color(0.5, 0.5, 0.5);
Color Color::Black = Color(0.0, 0.0, 0.0);
Color Color::Background = Color::Black;
Color Color::DefaultColor = Color::Black;

struct Camera
{
    Vector forward;
    Vector right;
    Vector up;
    Vector pos;

    Camera() {}
    Camera(Vector pos, Vector lookAt)
    {
        Vector Down = Vector(0.0, -1.0, 0.0);
        Vector Forward = lookAt - pos;
        this->pos = pos;
        this->forward = Forward.Norm();
        this->right = this->forward.Cross(Down).Norm() * 1.5;
        this->up = this->forward.Cross(this->right).Norm() * 1.5;
    }

    Vector GetPoint(int x, int y, int screenWidth, int screenHeight) const
    {
        double recenterX = (x - (screenWidth / 2.0)) / 2.0 / screenWidth;
        double recenterY = -(y - (screenHeight / 2.0)) / 2.0 / screenHeight;
        return (this->forward + ((this->right * recenterX) + (this->up * recenterY))).Norm();
    }
};

struct Ray
{
    Vector start;
    Vector dir;

    Ray(Vector start, Vector dir) : start(start), dir(dir) { }
};

struct Thing;

struct Intersection
{
    const Thing* thing;
    Ray ray;
    double dist;

    Intersection(const Thing* thing, Ray ray, double dist) :
        thing(thing), ray(ray), dist(dist)
    {}
};

struct SurfacePropreties
{
    Color Diffuse;
    Color Specular;
    double Reflect = 0.0;
    double Roughness = 0.0;

    SurfacePropreties() {}
    SurfacePropreties(Color diffuse, Color specular, double reflect, double roughness) :
        Diffuse(diffuse), Specular(specular), Reflect(reflect), Roughness(roughness)
    {}
};

struct Surface
{
    virtual SurfacePropreties GetSurfaceProperties(const Vector& pos) const = 0;
};

struct Light
{
    Vector pos;
    Color color;
    Light(Vector pos, Color color) : pos(pos), color(color) { }
};

struct Thing
{
    virtual Vector GetNormal(const Vector& pos) const = 0;
    virtual std::optional<Intersection> GetIntersection(const Ray& ray) const = 0;
    virtual Surface& GetSurface() const = 0;
    virtual ~Thing() = default;
};

class Sphere : public Thing {
    Surface& surface;
    Vector   center;
    double   radius2;
public:
    Sphere(Vector center, double radius, Surface& surface) : surface(surface), center(center), radius2(radius* radius) {}

    Vector GetNormal(const Vector& pos) const override {
        return (pos - center).Norm();
    }

    std::optional<Intersection> GetIntersection(const Ray& ray) const override {
        Vector eo = center - ray.start;
        double v = eo * ray.dir;
        if (v >= 0.0) {
            double disc = radius2 - ((eo * eo) - (v * v));
            if (disc >= 0.0) {
                double dist = v - sqrt(disc);
                return Intersection(this, ray, dist);
            }
        }
        return std::nullopt;
    }

    Surface& GetSurface() const override { return surface; };
};

class Plane : public Thing {
    Surface& surface;
    Vector   normal;
    double   offset;
public:
    Plane(Vector normal, double offset, Surface& surface) : surface(surface), normal(normal), offset(offset) {}

    Vector GetNormal(const Vector& pos) const override {
        return normal;
    }

    std::optional<Intersection> GetIntersection(const Ray& ray) const override {
        double denom = normal * ray.dir;
        if (denom > 0.0) {
            return std::nullopt;
        }
        double dist = ((normal * ray.start) + offset) / (-denom);
        return Intersection(this, ray, dist);
    }

    Surface& GetSurface() const override { return surface; };
};

struct ShinySurface : public Surface
{
    SurfacePropreties GetSurfaceProperties(const Vector& pos) const override
    {
        return SurfacePropreties(Color::White, Color::Grey, 0.7, 250.0);
    }
};

struct CheckerboardSurface : public Surface
{
    SurfacePropreties GetSurfaceProperties(const Vector& pos) const override
    {
        Color diffuse = Color::Black;
        double reflect = 0.7;
        if (((int)(floor(pos.z) + floor(pos.x))) % 2 != 0)
        {
            diffuse = Color::White;
            reflect = 0.1;
        }
        return SurfacePropreties(diffuse, Color::White, reflect, 150.0);
    }
};

class Scene {
private:
    ShinySurface        shiny;
    CheckerboardSurface checkerboard;
public:
    std::vector<std::unique_ptr<Thing>> things;
    std::vector<Light> lights;
    Camera    camera;

    Scene()
    {
        things.push_back(std::make_unique<Plane>(Vector(0.0, 1.0, 0.0), 0.0, checkerboard));
        things.push_back(std::make_unique<Sphere>(Vector(0.0, 1.0, -0.25), 1.0, shiny));
        things.push_back(std::make_unique<Sphere>(Vector(-1.0, 0.5, 1.5), 0.5, shiny));

        lights.push_back(Light(Vector(-2.0, 2.5, 0.0), Color(0.49, 0.07, 0.07)));
        lights.push_back(Light(Vector(1.5, 2.5, 1.5), Color(0.07, 0.07, 0.49)));
        lights.push_back(Light(Vector(1.5, 2.5, -1.5), Color(0.07, 0.49, 0.071)));
        lights.push_back(Light(Vector(0.0, 3.5, 0.0), Color(0.21, 0.21, 0.35)));
        camera = Camera(Vector(3.0, 2.0, 4.0), Vector(-1.0, 0.5, 0.0));
    }
};

class RayTracerEngine
{
    static const int maxDepth = 5;
    Scene& scene;

    std::optional<Intersection> GetClosestIntersection(const Ray& ray)
    {
        double closest = FarAway;
        std::optional<Intersection> closestInter = std::nullopt;

        for (auto& thing : scene.things)
        {
            auto inter = thing->GetIntersection(ray);
            if (inter && inter->dist < closest) {
                closestInter = inter;
                closest = inter->dist;
            }
        }
        return closestInter;
    }

    Color TraceRay(const Ray& ray, int depth)
    {
        auto isect = GetClosestIntersection(ray);
        if (isect) {
            return Shade(*isect, depth);
        }
        return Color::Background;
    }

    Color Shade(const Intersection& isect, int depth)
    {
        Vector d = isect.ray.dir;
        Vector pos = (d * isect.dist) + isect.ray.start;
        Vector normal = isect.thing->GetNormal(pos);
        Vector reflectDir = (d - ((normal * (normal * d)) * 2)).Norm();

        SurfacePropreties surface = isect.thing->GetSurface().GetSurfaceProperties(pos);

        Color naturalColor = Color::Background + GetNaturalColor(surface, pos, normal, reflectDir);
        Color reflectedColor = (depth >= maxDepth) ? Color::Grey : GetReflectionColor(surface, pos, reflectDir, depth);

        return naturalColor + reflectedColor;
    }

    Color GetReflectionColor(const SurfacePropreties& surface, const Vector& pos, const Vector& reflectDir, int depth)
    {
        Ray    ray(pos, reflectDir);
        Color  color = TraceRay(ray, depth + 1);
        return color.Scale(surface.Reflect);
    }

    Color GetNaturalColor(const SurfacePropreties& surface, const Vector& pos, const Vector& norm, const Vector& reflectDir)
    {
        Color result = Color::Black;
        for (auto& light : scene.lights)
        {
            Vector ldis = light.pos - pos;
            Vector livec = ldis.Norm();
            Ray ray{ pos, livec };

            auto neatIsect = GetClosestIntersection(ray);
            bool isInShadow = neatIsect.has_value() ? (neatIsect->dist <= ldis.Length()) : false;

            if (!isInShadow) {
                double illum = livec * norm;
                double specular = livec * reflectDir;

                Color lcolor = (illum > 0) ? (light.color.Scale(illum)) : Color::DefaultColor;
                Color scolor = (specular > 0) ? (light.color.Scale(pow(specular, surface.Roughness))) : Color::DefaultColor;
                result = result + lcolor * surface.Diffuse + scolor * surface.Specular;
            }
        }
        return result;
    }

public:
    RayTracerEngine(Scene& scene) : scene(scene) {}

    void render(RgbColor* image, int w, int h)
    {
        Ray ray(scene.camera.pos, Vector());
        auto& camera = scene.camera;
        int pos = 0;

        for (int y = 0; y < h; ++y) {
            pos = y * h;
            for (int x = 0; x < w; ++x) {
                ray.dir = camera.GetPoint(x, y, w, h);
                image[pos + x] = TraceRay(ray, 0).ToDrawingColor();
            }
        }
    }
};

void SaveImage(RgbColor* bitmapBits, int width, int height, const char* fileName)
{
    typedef unsigned int DWORD;
    typedef int LONG;
    typedef unsigned short WORD;
    const int BI_RGB = 0;

    struct BITMAPINFOHEADER {
        DWORD biSize;
        LONG  biWidth;
        LONG  biHeight;
        WORD  biPlanes;
        WORD  biBitCount;
        DWORD biCompression;
        DWORD biSizeImage;
        LONG  biXPelsPerMeter;
        LONG  biYPelsPerMeter;
        DWORD biClrUsed;
        DWORD biClrImportant;
    };

#pragma pack(push, 1)
    struct BITMAPFILEHEADER {
        WORD  bfType;
        DWORD bfSize;
        WORD  bfReserved1;
        WORD  bfReserved2;
        DWORD bfOffBits;
    };
#pragma pack(pop)

    BITMAPINFOHEADER bmpInfoHeader = { 0 };
    bmpInfoHeader.biSize = sizeof(BITMAPINFOHEADER);
    bmpInfoHeader.biBitCount = 32;
    bmpInfoHeader.biClrImportant = 0;
    bmpInfoHeader.biClrUsed = 0;
    bmpInfoHeader.biCompression = BI_RGB;
    bmpInfoHeader.biHeight = -height;
    bmpInfoHeader.biWidth = width;
    bmpInfoHeader.biPlanes = 1;
    bmpInfoHeader.biSizeImage = width * height * 4;

    BITMAPFILEHEADER bfh = { 0 };
    bfh.bfType = 'B' + ('M' << 8);
    bfh.bfOffBits = sizeof(BITMAPINFOHEADER) + sizeof(BITMAPFILEHEADER);
    bfh.bfSize = bfh.bfOffBits + bmpInfoHeader.biSizeImage;

    std::ofstream file(fileName, std::ios::binary | std::ios::trunc);
    file.write((const char*)&bfh, sizeof(bfh));
    file.write((const char*)&bmpInfoHeader, sizeof(bmpInfoHeader));
    file.write((const char*)bitmapBits, bmpInfoHeader.biSizeImage);
    file.close();
}

int main()
{
    std::cout << "Started " << std::endl;
    auto t1 = std::chrono::high_resolution_clock::now();

    Scene scene;
    RayTracerEngine rayTracer(scene);

    const int width = 500;
    const int height = 500;

    std::vector<RgbColor> bitmapData(width * height);
    rayTracer.render(&bitmapData[0], width, height);

    auto t2 = std::chrono::high_resolution_clock::now();
    auto diff = std::chrono::duration_cast<std::chrono::milliseconds>((t2 - t1));

    std::cout << "Completed in " << diff.count() << " ms" << std::endl;
    SaveImage(&bitmapData[0], width, height, "cpp-raytracer.bmp");

    return 0;
};

File: nim/RayTracer.nim
import math
import times
import macros
type
  SurfaceType = enum
    ShinySurface, CheckerBoardSurface

  ObjectType = enum
    Sphere, Plane

  RgbColor {.packed.} = object
    b, g, r, a: uint8

  Vector = object
    x, y, z: float64

  Color = object
    r, g, b: float64

  VC = Vector | Color

  Camera = object
    forward, right, up, pos: Vector

  Ray = object
    start, dir: Vector

  Thing = object
    surfaceType: SurfaceType
    case objectType: ObjectType
    of Sphere:
      center: Vector
      radius2: float64
    of Plane:
      normal: Vector
      offset: float64

  Intersection = object
    thingidx: int
    ray: Ray
    dist: float64

  Light = object
    pos: Vector
    color: Color

  Scene[T, L: static int] = object
    maxDepth: int
    things: array[T, Thing]
    lights: array[L, Light]
    camera: Camera

  SurfaceProperties = object
    diffuse, specular: Color
    reflect: float64
    roughness: uint

#fast pow
#[func pow(x: float64, y: uint): float64 =
  result = 1.0
  var
    base = x
    exp = y
  while exp != 0:
    if (exp and 1) == 1:
      result *= base
    exp = exp div 2
    base *= base
  ]#
func pow(x: float64, y: uint): float64 = pow(x, y.float64)
#init procs
func initVector(x, y, z: float64): Vector{.noinit, inline.} =
  result.x = x
  result.y = y
  result.z = z
func initColor(r, g, b: float64): Color{.noinit, inline.} =
  result.r = r
  result.g = g
  result.b = b
func initRay(s, d: Vector): Ray{.noinit, inline.} =
  result.start = s
  result.dir = d

func toRgbColor(c: Color): RgbColor{.noinit, inline.} =
  #as long as we never have a negative color
  template legalize(c: float64): uint8 = (if c > 1.0: 255'u8 else: uint8(c*255.0))
  result.r = legalize(c.r)
  result.g = legalize(c.g)
  result.b = legalize(c.b)
  result.a = 255

#wankery i cant help
macro getfield(x: untyped, f: static string): untyped =
  let id = ident(f)
  result = quote do:
    `x`.`id`
template fieldop(res, obj1, obj2: typed, op: untyped): untyped =
  for fld, val1, val2 in fieldpairs(obj1, obj2):
    let
      a{.inject.} = obj1.getfield(fld)
      b{.inject.} = obj2.getfield(fld)
    res.getfield(fld) = op

###Vector Math
template cross(v1, v2: Vector): Vector =
  initVector(
    x = v1.y * v2.z - v1.z * v2.y,
    y = v1.z * v2.x - v1.x * v2.z,
    z = v1.x * v2.y - v1.y * v2.x
  )

template len(v: Vector): float64 =
  sqrt(v.x * v.x + v.y * v.y + v.z * v.z)

func `*`[T: VC](v: T, k: float64): T =
  for fld, val in v.fieldpairs:
    result.getfield(fld) = v.getfield(fld) * k

func `*`[T: VC](v1, v2: T): T = fieldop(result, v1, v2, a*b)
func `+`[T: VC](v1, v2: T): T = fieldop(result, v1, v2, a+b)
func `-`[T: VC](v1, v2: T): T = fieldop(result, v1, v2, a - b)
func `+=`[T: VC](v1: var T, v2: T) = fieldop(v1, v1, v2, a+b)
template norm(v: Vector): Vector =
  v * (1.0 / v.len)

func dot(v1, v2: Vector): float64{.inline.} =
  (v1.x * v2.x) + (v1.y * v2.y) + (v1.z * v2.z)

func getnormal(obj: Thing, pos: Vector): Vector{.inline.} =
  case obj.objectType:
    of Sphere:
      (pos - (obj.center)).norm()
    of Plane:
      obj.normal

#setup
const
  white = Color(r: 1.0, g: 1.0, b: 1.0)
  grey = Color(r: 0.5, g: 0.5, b: 0.5)
  black = Color(r: 0.0, g: 0.0, b: 0.0)
  background = black
  defaultColor = black


func initCamera(pos: Vector, lookAt: Vector): Camera =
  let down = Vector(x: 0.0, y: -1.0, z: 0.0)
  let forward = lookAt - pos

  var camera: Camera
  camera.pos = pos
  camera.forward = forward.norm()
  camera.right = camera.forward.cross(down)
  camera.up = camera.forward.cross(camera.right)

  let rightNorm = camera.right.norm()
  let upNorm = camera.up.norm()

  camera.right = rightNorm * 1.5
  camera.up = upNorm * 1.5

  return camera

proc initSphere(center: Vector, radius: float64,
    surfaceType: SurfaceType): Thing =
  var r2 = radius * radius
  return Thing(surfaceType: surfaceType, objectType: Sphere, center: center, radius2: r2)

proc initPlane(normal: Vector, offset: float64,
    surfaceType: SurfaceType): Thing =
  return Thing(surfaceType: surfaceType, objectType: Plane, normal: normal,
      offset: offset)

func initScene(): auto =
  let
    maxDepth = 5

    things = [
      initPlane(Vector(x: 0.0, y: 1.0, z: 0.0), 0.0, CheckerBoardSurface),
      initSphere(Vector(x: -1.0, y: 0.5, z: 1.5), 0.5, ShinySurface),
      initSphere(Vector(x: 0.0, y: 1.0, z: -0.25), 1.0, ShinySurface)
      ]

    lights = [
      Light(pos: Vector(x: -2.0, y: 2.5, z: 0.0), color: Color(r: 0.49,
        g: 0.07, b: 0.07)),
      Light(pos: Vector(x: 1.5, y: 2.5, z: 1.5), color: Color(r: 0.07,
        g: 0.07, b: 0.49)),
      Light(pos: Vector(x: 1.5, y: 2.5, z: -1.5), color: Color(r: 0.07,
        g: 0.49, b: 0.071)),
      Light(pos: Vector(x: 0.0, y: 3.5, z: 0.0), color: Color(r: 0.21,
        g: 0.21, b: 0.35))
    ]

    camera = initCamera(Vector(x: 3.0, y: 2.0, z: 4.0), Vector(x: -1.0,
      y: 0.5, z: 0.0))
  return Scene[things.len, lights.len](maxDepth: maxDepth, things: things,
      lights: lights, camera: camera)



##   the real meat of the program ##
##
const shiny = SurfaceProperties(diffuse: white, specular: grey, reflect: 0.7,
    roughness: 250)
func getSurfaceProperties(obj: Thing, pos: Vector): SurfaceProperties{.noinit.} =
  case obj.surfaceType:
  of ShinySurface: result = shiny
  of CheckerBoardSurface:
    let val = (int)(floor(pos.z) + floor(pos.x)) mod 2
    if val == 0:
      result.reflect = 0.7
      result.diffuse = black
    else:
      result.reflect = 0.1
      result.diffuse = white
    result.specular = white
    result.roughness = 150

template intersections(obj: Thing, start, dir: Vector, test,
    body: untyped): untyped =
  var dist{.inject.}: float64
  case obj.objectType:
  of Sphere:
    let
      eo = obj.center - start
      v = eo.dot(dir)
    if v >= 0:
      let disc = obj.radius2 - (eo.dot(eo) - (v * v))
      if disc >= 0:
        dist = v - sqrt(disc)
        if test:
          body
  of Plane:
    let denom = obj.normal.dot(dir)
    if (denom <= 0) and test:
      dist = (obj.normal.dot(start) + obj.offset) / (-denom)
      body
func getNaturalColor(scene: Scene, sp: SurfaceProperties,
    pos, rd, norm: Vector): Color

func getReflectionColor(scene: var Scene, sp: SurfaceProperties, pos,
    rd: Vector, depth: int): Color

func shade(scene: var Scene, thing: Thing, start, dir: Vector, dist: float64,
    depth: int): Color =
  let
    scaled = dir * dist
    pos = scaled + start
    #thing = scene.things[thingidx]
    normal = thing.getnormal(pos)
    reflectDir = dir - (normal * (normal.dot(dir) * 2))
    sp = thing.getSurfaceProperties(pos)
  when background == black:
    let naturalColor = getNaturalColor(scene, sp, pos, reflectDir, normal)
  else:
    let naturalColor = getNaturalColor(scene, sp, pos, reflectDir, normal) + background

  let reflectedColor = if depth >= scene.maxDepth:
      grey
    else:
      getReflectionColor(scene, sp, pos, reflectDir, depth)

  return naturalColor+reflectedColor



func testRay(scene: Scene, start, dir: Vector): float64 =
  result = INF
  for thing in scene.things:
    intersections(thing, start, dir, dist < result):
      result = dist
func traceRay(scene: var Scene, start, dir: Vector, depth: int): Color =
  var
    closest = INF
    closestThing: ptr Thing
  for thng in scene.things.mitems:
    intersections(thng, start, dir, dist < closest):
      closest = dist
      closestThing = thng.addr
  result = if closestThing != nil:
    scene.shade(closestThing[], start, dir, closest, depth)
  else:
    black
#ray = Ray(start:pos,dir:rd)
#props = getsurfaceprops(thing,pos)
func getReflectionColor(scene: var Scene, sp: SurfaceProperties, pos,
    rd: Vector, depth: int): Color =
  scene.traceRay(pos, rd, depth + 1) * sp.reflect

func getNaturalColor(scene: Scene, sp: SurfaceProperties,
    pos, rd, norm: Vector): Color =
  var rdNorm = rd.norm()

  for light in scene.lights:
    let
      ldis = light.pos - pos
      ldisLen = ldis.len()
      livec = ldis.norm() #ldis / ldisLen) #ldis.norm
                          #lray = Ray(start: pos, dir: livec)
      neatIsect = scene.testRay(pos, livec)

    if neatIsect > ldisLen:
      let illum = livec.dot(norm)
      let specular = livec.dot(rdNorm)

      #[let lcolor = if illum > 0:
        light.color * illum * sp.diffuse
      else:
        defaultColor]#
      assert illum > 0.0
      let lcolor = (light.color * illum) * sp.diffuse
      let scolor = if specular > 0:
                     light.color * pow(specular, sp.roughness) * sp.specular
                   else:
                     defaultColor

      result += lcolor + scolor



#[proc ObjectIntersect(obj: Thing, ray: Ray): Intersection =
  case obj.objectType:
    of Sphere:
      let eo = obj.center.Sub(ray.start)
      let v = eo.Dot(ray.dir)
      var dist = 0.0
      if (v >= 0):
        let disc = obj.radius2 - (eo.Dot(eo) - (v * v))
        if (disc >= 0):
          dist = v - sqrt(disc)
        if (dist != 0.0):
          result.thing = obj
          result.ray = ray
          result.dist = dist
        of Plane:
          let denom = obj.normal.Dot(ray.dir)
          if (denom <= 0):
            result.dist = (obj.normal.Dot(ray.start) + obj.offset) / (-denom)
            result.thing = obj
            result.ray = ray

          proc Intersections(scene: Scene, ray: Ray): Intersection =
            var closest: float64 = FarAway
            result.thing = nil

            for thing in scene.things:
              let intersect = ObjectIntersect(thing, ray)
              if (not isNil(intersect.thing)) and (intersect.dist < closest):
            result = intersect
            closest = intersect.dist
              return result
            ]#


### Render Scene
template getPoint(x, y: int, camera: Camera, sw, sh: float64): Vector =
  let
    recenterX = (x.float64 - (sw / 2.0)) / (2.0 * sw)
    recenterY = -(y.float64 - (sh / 2.0)) / (2.0 * sh)

    vx = camera.right * recenterX
    vy = camera.up * recenterY

    v = vx + vy
    z = camera.forward + v

  z.norm()

proc RenderScene(scene: var Scene, bitmapData: var seq[RgbColor], stride: int,
    w: int, h: int) =
  let start = scene.camera.pos
  var dir: Vector
  let
    wf = w.float64
    hf = h.float64
  for y in 0 ..< h:
    var pos = y * w
    for x in 0 ..< w:
      dir = getPoint(x, y, scene.camera, wf, hf)
      bitmapData[pos] = scene.traceRay(start, dir, 0).toRgbColor()
      pos = pos + 1


#Bitmap
proc SaveRGBBitmap(bitmapData: seq[RgbColor], width: int, height: int,
    wBitsPerPixel: int, fileName: string) =
  type DWORD = uint32
  type LONG = int32
  type WORD = int16
  const BI_RGB = 0

  type BITMAPINFOHEADER {.packed.} = object
    biSize: DWORD
    biWidth: LONG
    biHeight: LONG
    biPlanes: WORD
    biBitCount: WORD
    biCompression: DWORD
    biSizeImage: DWORD
    biXPelsPerMeter: LONG
    biYPelsPerMeter: LONG
    biClrUsed: DWORD
    biClrImportant: DWORD

  type BITMAPFILEHEADER {.packed.} = object
    bfType: WORD
    bfSize: DWORD
    bfReserved1: WORD
    bfReserved2: WORD
    bfOffBits: DWORD

  var bmpInfoHeader: BITMAPINFOHEADER
  bmpInfoHeader.biSize = DWORD(sizeof(BITMAPINFOHEADER))
  bmpInfoHeader.biBitCount = WORD(wBitsPerPixel)
  bmpInfoHeader.biClrImportant = 0
  bmpInfoHeader.biClrUsed = 0
  bmpInfoHeader.biCompression = BI_RGB
  bmpInfoHeader.biHeight = LONG(-height)
  bmpInfoHeader.biWidth = LONG(width)
  bmpInfoHeader.biPlanes = 1
  bmpInfoHeader.biSizeImage = DWORD(width * height * (wBitsPerPixel div 8))

  var bfh: BITMAPFILEHEADER
  bfh.bfType = 0x4D42 #'B' + ('M'shl 8)
  bfh.bfOffBits = sizeof(BITMAPINFOHEADER) + sizeof(BITMAPFILEHEADER)
  bfh.bfSize = bfh.bfOffBits + bmpInfoHeader.biSizeImage

  var file = open(fileName, fmWrite)

  discard file.writeBuffer(addr bfh, sizeof(BITMAPFILEHEADER))
  discard file.writeBuffer(addr bmpInfoHeader, sizeof(BITMAPINFOHEADER))
  for c in bitmapData:
    var x = c
    discard file.writeBuffer(addr x, sizeof(RgbColor))
  file.close()


var t1 = cpuTime()
var scene = initScene()
var width = 500
var height = 500
var stride = width * 4
var bitmapData = newSeq[RgbColor](width * height)

RenderScene(scene, bitmapData, stride, width, height)
var t2 = cpuTime()
var diff = (t2 - t1) * 1000

echo "CPU time [ms] ", diff

SaveRGBBitmap(bitmapData, width, height, 32, "nim-raytracer.bmp")
